Documentation revamp (#4871)
spotify/scio: site/src/main/paradox/extras/AsyncDoFn.md

> # AsyncDoFn

Reviewer: Thank you :heart:

Scio's @scaladoc[`BaseAsyncDoFn`](com.spotify.scio.transforms.BaseAsyncDoFn) provides standard handling for sending asynchronous requests and capturing the responses for a bundle of pipeline elements.
`BaseAsyncDoFn` is a subclass of @scaladoc[`DoFnWithResource`](com.spotify.scio.transforms.DoFnWithResource) which handles the creation and re-use of client classes.
Scio provides several future-specific subclasses to choose from depending on the return type of the client:

* @scaladoc[`GuavaAsyncDoFn`](com.spotify.scio.transforms.GuavaAsyncDoFn) for clients that return Guava's `ListenableFuture`
* @scaladoc[`JavaAsyncDoFn`](com.spotify.scio.transforms.JavaAsyncDoFn) for clients that return `CompletableFuture`
* @scaladoc[`ScalaAsyncDoFn`](com.spotify.scio.transforms.ScalaAsyncDoFn) for clients that return a scala `Future`

`BaseAsyncDoFn` will wait for all futures for all bundle elements to be returned before completing the bundle.

> A failure of any request for an item in the bundle will cause the entire bundle to be retried.

Reviewer 2: ```suggestion
A failure of any request for an item in the bundle will cause the entire bundle to be retried; take care to make your GRPC requests idempotent, as there's no guarantee they'll be called exactly once.
```

> Given this mock client:

Reviewer 2: ```suggestion
Given this Guava-based mock client:
```

```scala
import com.google.common.util.concurrent.{ListenableFuture, Futures}

case class MyClient(value: String) {
  def request(i: Int): ListenableFuture[String] = Futures.immediateFuture(s"$value$i")
}
```

> A custom `DoFn` can be defined using `GuavaAsyncDoFn`, since the client returns `ListenableFuture`.

Reviewer 2: ```suggestion
We can invoke it in a Scio pipeline by implementing a `GuavaAsyncDoFn`, which supports the `ListenableFuture` interface.
```

Author: Trying to avoid "we" and "you" as a style preference, though sometimes unavoidable
